src/lib/__tests__/index-test.ts

```diff
@@ -0,0 +1,855 @@
+
+      expect(() => {
+        // @ts-ignore
+        instance.addWidgets();
+      }).toThrow();
+
+      expect(() => {
+        instance.addWidgets(createWidget() as any);
+      }).toThrowErrorMatchingInlineSnapshot(`
+"The \`addWidgets\` method expects an array of widgets.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+`);
+    });
+
+    it('throws an error with widgets that do not implement `init` or `render`', () => {
+      const instance = index({ indexName: 'index_name' });
+
+      expect(() => {
+        instance.addWidgets([{ dummy: true } as any]);
+      }).toThrowErrorMatchingInlineSnapshot(`
+"The widget definition expects a \`render\` and/or an \`init\` method.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+`);
+    });
+
+    describe('with a started instance', () => {
+      it('updates the internal state with added widgets', () => {
+        const instance = index({ indexName: 'index_name' });
+        const instantSearchInstance = createInstantSearch();
+
+        instance.addWidgets([createSearchBox()]);
+
+        instance.init!(
+          createInitOptions({
+            instantSearchInstance,
+          })
+        );
+
+        expect(instance.getHelper()!.state).toEqual(
+          algoliasearchHelper.SearchParameters.make({
+            index: 'index_name',
+            query: 'Apple',
+          })
+        );
+
+        instance.addWidgets([createPagination()]);
+
+        expect(instance.getHelper()!.state).toEqual(
+          algoliasearchHelper.SearchParameters.make({
+            index: 'index_name',
+            query: 'Apple',
+            page: 5,
+          })
+        );
+      });
+
+      it('calls `init` on the added widgets', () => {
+        const instance = index({ indexName: 'index_name' });
+        const instantSearchInstance = createInstantSearch();
+        const pagination = createSearchBox({
+          init: jest.fn(),
+        });
+
+        instance.init!(
+          createInitOptions({
+            instantSearchInstance,
+          })
+        );
+
+        expect(pagination.init).toHaveBeenCalledTimes(0);
+
+        instance.addWidgets([pagination]);
+
+        expect(pagination.init).toHaveBeenCalledTimes(1);
+        expect(pagination.init).toHaveBeenCalledWith({
+          instantSearchInstance,
+          parent: instance,
+          helper: instance.getHelper(),
+          state: instance.getHelper()!.state,
+          templatesConfig: instantSearchInstance.templatesConfig,
+          createURL: instantSearchInstance._createAbsoluteURL,
+          onHistoryChange: instantSearchInstance._onHistoryChange,
+        });
+      });
+
+      it('schedules a search to take the added widgets into account', () => {
+        const instance = index({ indexName: 'index_name' });
+        const instantSearchInstance = createInstantSearch({
+          scheduleSearch: jest.fn(),
```

Comment:
Those tests are brittle, but the method doesn't exist yet on InstantSearch. We might want to actually use the instance when it's available. We won't "unit" test anymore but at least we'll test a real behaviour.